mmp: Show a nice error if user tries to build a 32-bit app with Xcode 10

Repro: build a 32-bit XM app with Xcode 10

Result: many linker failures (full sample log: https://gist.github.com/mykyta-bondarenko-gl/3e7b8c1cee9fcfc546c491c79a12faa0#file-output-L1411)

    Undefined symbols for architecture i386:
      ".objc_class_name_NSAlert", referenced from:
          pointer-to-literal-objc-class-name in libxammac-debug.a(launcher-debug.static.i386.o)
      ".objc_class_name_NSArray", referenced from:
          pointer-to-literal-objc-class-name in libxammac-debug.a(trampolines-debug.static.i386.o)
          pointer-to-literal-objc-class-name in libxammac-debug.a(trampolines-invoke-debug.static.i386.o)
      ".objc_class_name_NSAutoreleasePool", referenced from:
          pointer-to-literal-objc-class-name in libxammac-debug.a(shared-debug.static.i386.o)
      ".objc_class_name_NSBundle", referenced from:
          pointer-to-literal-objc-class-name in libxammac-debug.a(runtime-debug.static.i386.o)
          pointer-to-literal-objc-class-name in libxammac-debug.a(launcher-debug.static.i386.o)
      ".objc_class_name_NSDictionary", referenced from:
          pointer-to-literal-objc-class-name in libxammac-debug.a(runtime-debug.static.i386.o)
          pointer-to-literal-objc-class-name in libxammac-debug.a(launcher-debug.static.i386.o)

We should report a better MM#### error instead.

--- a/docs/website/mmp-errors.md
+++ b/docs/website/mmp-errors.md
@@ -235,6 +235,18 @@ Alternatively, enable the managed [linker](https://docs.microsoft.com/xamarin/ma
 
 As a last-straw solution, use an older version of Xamarin.Mac that does not require these new SDKs to be present during the build process.
 
+#### MM0138: Building 32-bit apps is not possible when using Xcode 10. Please migrate project to the Unified API.
+
+Xcode 10 does not support building 32-bit applications.
+
+The project must be [migrated to a Unified project](https://docs.microsoft.com/en-us/xamarin/cross-platform/macios/unified/updating-mac-apps) in order to support 64-bit.
+
+#### MM0139: Building 32-bit apps is not possible when using Xcode 10. Please change the architecture in the project's Mac Build options to 'x86_64'.
+
+Xcode 10 does not support building 32-bit applications.
+
+Change the architecture in the project's Mac Build options to 'x86_64' in order to build a 64-bit application.
+
 ## MM1xxx: file copy / symlinks (project related)
 
 <a name="MM1034" />

--- a/docs/website/mtouch-errors.md
+++ b/docs/website/mtouch-errors.md
@@ -872,6 +872,8 @@ Alternatively, enable the managed [linker](https://docs.microsoft.com/en-us/xama
 
 As a last-straw solution, use an older version of Xamarin.iOS that does not require these new SDKs to be present during the build process.
 
+<!-- 0138-0139: used by mmp -->
+
 ## MT1xxx: Project related error messages
 
 ### MT10xx: Installer / mtouch

--- a/tests/mmptest/src/MMPTest.cs
+++ b/tests/mmptest/src/MMPTest.cs
@@ -650,5 +650,29 @@ namespace Xamarin.MMP.Tests
 				rv.Messages.AssertErrorCount (0);
 			});
 		}
+
+		[Test]
+		public void MM0138 ()
+		{
+			MMPTests.RunMMPTest (tmpDir => {
+				var rv = TI.TestClassicExecutable (tmpDir, csprojConfig: "<IncludeMonoRuntime>true</IncludeMonoRuntime>", shouldFail: true);
+				rv.Messages.AssertError (138, "Building 32-bit apps is not possible when using Xcode 10. Please migrate project to the Unified API.");
+				rv.Messages.AssertWarningCount (0);
+			});
+		}
+
+		[Test]
+		public void MM0139 ()
+		{
+			RunMMPTest (tmpDir => {
+				TI.UnifiedTestConfig test = new TI.UnifiedTestConfig (tmpDir) {
+					CSProjConfig = "<XamMacArch>i386</XamMacArch>"
+				};
+
+				var rv = TI.TestUnifiedExecutable (test, shouldFail: true);
+				rv.Messages.AssertError (139, "Building 32-bit apps is not possible when using Xcode 10. Please change the architecture in the project's Mac Build options to 'x86_64'.");
+				rv.Messages.AssertWarningCount (0);
+			});
+		}
 	}
 }

--- a/tools/mmp/driver.cs
+++ b/tools/mmp/driver.cs
@@ -484,6 +484,12 @@ namespace Xamarin.Bundler {
 			// InitializeCommon needs SdkVersion set to something valid
 			ValidateSDKVersion ();
 
+			if (action != Action.RunRegistrar && XcodeVersion.Major >= 10 && arch != "x86_64") {
+				if (IsClassic)
+					throw ErrorHelper.CreateError (138, "Building 32-bit apps is not possible when using Xcode 10. Please migrate project to the Unified API.");
+				throw ErrorHelper.CreateError (139, "Building 32-bit apps is not possible when using Xcode 10. Please change the architecture in the project's Mac Build options to 'x86_64'.");
+			}
+
 			// InitializeCommon needs the current profile
 			if (IsClassic)
 				Profile.Current = new MonoMacProfile ();

--- a/tools/mmp/driver.cs
+++ b/tools/mmp/driver.cs
@@ -484,7 +484,7 @@ namespace Xamarin.Bundler {
 			// InitializeCommon needs SdkVersion set to something valid
 			ValidateSDKVersion ();
 
-			if (action != Action.RunRegistrar && XcodeVersion.Major >= 10 && arch != "x86_64") {
+			if (action != Action.RunRegistrar && XcodeVersion.Major >= 10 && !Is64Bit) {
 				if (IsClassic)
 					throw ErrorHelper.CreateError (138, "Building 32-bit apps is not possible when using Xcode 10. Please migrate project to the Unified API.");
 				throw ErrorHelper.CreateError (139, "Building 32-bit apps is not possible when using Xcode 10. Please change the architecture in the project's Mac Build options to 'x86_64'.");

